### PR TITLE
281-Create All Therapists page

### PR DESCRIPTION
**Pull request by AdhamHaisami**
Relates #281

#### Issues this PR relates to

=======

_NOTE: REMEMBER NOT TO REVIEW UNLESS ALL ACCEPTANCE CRITERIA HAS BEEN TICKED OFF_

---

#### Any info the reviewer needs to know to test 
e.g. Add env vars, run npm install

---

#### Any info the team needs to consider before merged into development
e.g. Will a database need to be rebuilt (and is this ok if existing data), add env vars

---

#### Quick Pull Request Checklist

Please make sure you ensure the following has been done before marking your PR as awaiting-review. For full checklist and process review, please refer to Process Guide (#8)

- Branch is up to date with develop
- Project runs
- You can download and install locally, and run the project locally
- All existing and new features run without bugs
- Checked that features/design works across all agreed browsers + devices
- The new work has been checked with Axe accessibility checker and doesn't throw significant errors  
- Linter is used
- Code is consistent with agreed team style
- Naming of files is consistent with project 
- Tests all pass
- No console logs

### CODE REVIEW

File: server/modules/user/use-cases/update-user-role.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import * as User from '../model';
+
+const updateUserRole = async ({ id, role }) => {
```
Comment (RamyAlshurafa) [inline server/modules/user/use-cases/update-user-role.js:3]: add validation here, don't rely on the role from the client

File: server/database/models/users/schema.sql [hunk 2]
```diff
@@ -21,6 +21,7 @@ CREATE TABLE "users" (
   "organisation_id" INTEGER REFERENCES organisations(id) DEFAULT 1,
   "created_at" TIMESTAMP NOT NULL DEFAULT NOW(),
   "updated_at" TIMESTAMP NOT NULL DEFAULT NOW(),
+  "is_active" BOOLEAN DEFAULT TRUE,
```
Comment (RamyAlshurafa) [inline server/database/models/users/schema.sql:24]: we have status, so no need for this

File: client/src/pages/Admin/index.js [hunk 3]
```diff
@@ -52,6 +53,13 @@ function AdminRouter() {
             isPrivate
             allowedRoles={[roles.ADMIN]}
           />
+          <Route
+            path={navRoutes.ADMIN.All_THERAPISTS}
+            Component={AllTherapists}
+            layout="general"
+            isPrivate
+            allowedRoles={[roles.ADMIN]}
```
Comment (RamyAlshurafa) [inline client/src/pages/Admin/index.js:61]: >  This controller and page is authorised to be used / viewed by both admin and super admin

File: client/src/constants/roles.js [hunk 4]
```diff
@@ -4,4 +4,10 @@ const userRoles = {
   ADMIN: 'ADMIN',
 };
 
+export const therapistRoles = {
+  THERAPIST: 'THERAPIST',
+  ADMIN: 'ADMIN',
+  DEACTIVATED: 'DEACTIVATED',
+};
```
Comment (RamyAlshurafa) [inline client/src/constants/roles.js:11]: DEACTIVATED is status, not role

File: client/src/api-calls/users.js [hunk 5]
```diff
@@ -171,6 +171,28 @@ const getCSRFToken = async ({ options } = {}) => {
...
+      updates
+    );
+    return { data };
+  } catch (error) {
+    const err = handleError(error, options);
+    return { error: err };
+  }
+};
+
+const updateUserRole = async (updates, { options } = {}) => {
```
Comment (RamyAlshurafa) [inline client/src/api-calls/users.js:187]: i can see the api url is not related to roles, so I think the name should be updateUser,
- [ ] also the update should be /users/id

File: client/src/api-calls/users.js [hunk 6]
```diff
@@ -171,6 +171,28 @@ const getCSRFToken = async ({ options } = {}) => {
     return { error: err };
   }
 };
+const deActivateUser = async (updates, { options } = {}) => {
+  try {
+    const { data } = await axios.patch(
+      `${USERS_BASE}/deactivate-user`,
```
Comment (RamyAlshurafa) [inline client/src/api-calls/users.js:177]:
```suggestion
      `${USERS_BASE}/${id}/deactivate`,
```

File: server/modules/organisation/controllers/index.js [hunk 7]
```diff
@@ -4,10 +4,12 @@ import getOrganisationById from './get-organisation-by-id';
 import updateOrganisation from './update-organisation';
 
 import { authenticate } from '../../../api/middlewares';
+import getAllTherapists from './get-all-therapists';
 
 const router = Router();
 
 router.get('/:id', authenticate(), getOrganisationById);
+router.get('/:id/therapists', authenticate(), getAllTherapists);
```
Comment (RamyAlshurafa) [inline server/modules/organisation/controllers/index.js:12]: authorization

File: server/modules/user/controllers/index.js [hunk 8]
```diff
@@ -51,6 +53,13 @@ router.patch(
 );
 router.post('/profile', authenticate(), csrfProtection, createTherapistProfile);
 router.patch('/account', authenticate(), csrfProtection, updateAccount);
+router.patch(
+  '/deactivate-user',
+  csrfProtection,
+  authenticate(),
+  deactivateUser,
```
Comment (RamyAlshurafa) [inline server/modules/user/controllers/index.js:60]: authorisation

File: server/modules/user/controllers/update-user.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import * as User from '../use-cases';
+
+const updateUser = async (req, res, next) => {
+  try {
+    const { id, role } = req.body;
+    const updated = await User.updateUserRole({
+      id,
+      role: [role],
```
Comment (RamyAlshurafa) [inline server/modules/user/controllers/update-user.js:8]: send it as single role, the usecase should handle it

File: server/modules/user/use-cases/login.js [hunk 10]
```diff
@@ -33,6 +39,13 @@ const login = async ({ email, password, reToken }) => {
         id: userWithSameEmail.id,
       });
       userWithSameEmail.hasProfile = hasProfile;
+    } else if (userWithSameEmail.roles[0] === userRoles.CLIENT) {
+      const myTherapist = await TherapistClient.findTherapistByClientId(
```
Comment (RamyAlshurafa) [inline server/modules/user/use-cases/login.js:43]: add the same logic for getUserInfo controller

**Review by RamyAlshurafa — CHANGES_REQUESTED** [review]

File: client/src/pages/Admin/index.js (hunk 3 above)
Comment (AdhamHaisami) [reply client/src/pages/Admin/index.js:61]: done

File: server/database/models/users/schema.sql (hunk 2 above)
Comment (AdhamHaisami) [reply server/database/models/users/schema.sql]: done

File: client/src/constants/roles.js (hunk 4 above)
Comment (AdhamHaisami) [reply client/src/constants/roles.js:11]: done

File: client/src/api-calls/users.js (hunk 5 above)
Comment (AdhamHaisami) [reply client/src/api-calls/users.js:187]: done

File: client/src/api-calls/users.js (hunk 6 above)
Comment (AdhamHaisami) [reply client/src/api-calls/users.js:177]: done

File: server/modules/organisation/controllers/index.js (hunk 7 above)
Comment (AdhamHaisami) [reply server/modules/organisation/controllers/index.js:12]: done

File: server/modules/user/controllers/index.js (hunk 8 above)
Comment (AdhamHaisami) [reply server/modules/user/controllers/index.js:60]: done

File: server/modules/user/controllers/update-user.js (hunk 9 above)
Comment (AdhamHaisami) [reply server/modules/user/controllers/update-user.js]: done

File: server/modules/user/use-cases/update-user-role.js (hunk 1 above)
Comment (AdhamHaisami) [reply server/modules/user/use-cases/update-user-role.js:7]: done

File: server/database/models/users/schema.sql (hunk 2 above)
Comment (RamyAlshurafa) [reply server/database/models/users/schema.sql]: still used in
`const findTherapistByClientId = async (id) => {`

**Review by RamyAlshurafa — CHANGES_REQUESTED** [review]
- [x] can't login as admin
- [ ] show therapists for one organisations
- [ ] update page url /admin/organisations/:id/therapists
- [ ] update API endpoint /organisations/:id/therapists

**Review by RamyAlshurafa — APPROVED** [review]

**Pull request merged**